File: 2025/shinyoung/2025-07-06-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        if not head.next:
            return head

        prev=head
        curr=head.next

        reverse_curr=None
        reverse_next = ListNode(prev.val, None)
        
        while curr.next:
            reverse_curr = ListNode(curr.val, reverse_next)
            # reverse_prev = ListNode(0, reverse_curr)
            # reverse_curr = reverse_prev
            reverse_next = reverse_curr
            prev=prev.next
            curr=curr.next 
        reverse_curr = ListNode(curr.val, reverse_next)

        return reverse_curr


            

        

File: 2025/JuyoungMoon/week5/2025-07-06-reverse-linked-list.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 5: Linked List
// https://github.com/juyomo/leetcode-study

// LeetCode #206.
// https://leetcode.com/problems/reverse-linked-list/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) {
            return head;
        }

        ListNode* curr = head;
        ListNode* prev = nullptr;
        ListNode* next = head->next;

        while (curr != nullptr) {
            curr->next = prev;
            
            prev = curr;
            curr = next;
            if (curr != nullptr) {
                next = curr->next;
            }
        }
        
        return prev;
    }
};
